### DISCUSSION TITLE
How do I set up mitmproxy as HTTPS proxy?

### DISCUSSION THREAD

**Discussion by rootkea** [mitmproxy/mitmproxy#5317] [Q&A]
Hello!

`mitmproxy` serves at http://localhost:8080. How can I make it serve over `https` at https://localhost:8080?

My use-case is: I'm testing a http(s) client which can use `http` as well as `https` proxy to make connections. So how do I make `mitmproxy` act as a HTTPS proxy?

Thanks!

**Comment by mhils** [comment]
mitmproxy acts both as an HTTP and as an HTTPS proxy by default, we automatically detect if there's an additional TLS handshake first. You just need to tell your client it's HTTPS. :)

**Reply by rootkea** [reply]
Hello @mhils!

Thanks for your reply.

I tried:
```
$ curl --proxy https://127.0.0.1:8080 --proxy-cacert ~/.mitmproxy/mitmproxy-ca-cert.pem http://paulgraham.com
curl: (60) SSL: unable to obtain common name from peer certificate
More details here: https://curl.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
$
```

But I can't seem to talk `https`  with `mitmproxy` as explained by `curl` maintainer here: https://daniel.haxx.se/blog/2016/11/26/https-proxy-with-curl/

**Reply by rootkea** [reply]
I see "Common Name" in the cert though...

**Reply by mhils** [reply]
The exact command you posted works fine for me with just `mitmdump` (no args) running. Which mitmproxy version are you on?

**Reply by ghost** [reply]
@mhils I'm running into the same issue.
I'm running `mitmweb` with no arguments.
Using https for a proxy returns the following:

```
$ curl --cacert ~/.mitmproxy/mitmproxy-ca-cert.pem --proxy https://127.0.0.1:8080 https://google.com
curl: (60) SSL certificate problem: unable to get local issuer certificate
More details here: https://curl.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
```

I'm running mitmproxy 10.4.2 on Ubuntu 24.04

Using `mitmdump` prints the following
```
[12:13:52.992][127.0.0.1:51492] client connect
[12:13:53.008][127.0.0.1:51492] Client TLS handshake failed. The client does not trust the proxy's certificate for <no address> (tlsv1 alert unknown ca)
[12:13:53.009][127.0.0.1:51492] client disconnect
```

**Reply by Prinzhorn** [reply]
`--proxy-cacert ~/.mitmproxy/mitmproxy-ca-cert.pem `

**Reply by ghost** [reply]
This helped - thank you.
For clarification, the argument `--proxy-cacert` is expected in the `curl` command

```
curl \
  --cacert ~/.mitmproxy/mitmproxy-ca-cert.pem \
  --proxy-cacert ~/.mitmproxy/mitmproxy-ca-cert.pem \
  --proxy https://127.0.0.1:8080 \
  https://google.com
```
